src/USER-MDI/fix_mdi_engine.cpp

```diff
@@ -0,0 +1,932 @@
+/* ----------------------------------------------------------------------
+   LAMMPS - Large-scale Atomic/Molecular Massively Parallel Simulator
+   http://lammps.sandia.gov, Sandia National Laboratories
+   Steve Plimpton, sjplimp@sandia.gov
+
+   Copyright (2003) Sandia Corporation.  Under the terms of Contract
+   DE-AC04-94AL85000 with Sandia Corporation, the U.S. Government retains
+   certain rights in this software.  This software is distributed under
+   the GNU General Public License.
+
+   See the README file in the top-level LAMMPS directory.
+------------------------------------------------------------------------- */
+
+/* ----------------------------------------------------------------------
+   Contributing author: Taylor Barnes (MolSSI)
+   MolSSI Driver Interface (MDI) support for LAMMPS
+------------------------------------------------------------------------- */
+
+#include "fix_mdi_engine.h"
+#include "atom.h"
+#include "domain.h"
+#include "comm.h"
+#include "update.h"
+#include "force.h"
+#include "error.h"
+#include "group.h"
+#include "memory.h"
+#include "modify.h"
+#include "compute.h"
+
+#include <stdlib.h>
+#include <string.h>
+#include <limits>
+
+#include "irregular.h"
+#include "min.h"
+#include "minimize.h"
+#include "modify.h"
+#include "output.h"
+#include "timer.h"
+#include "verlet.h"
+#include "mdi.h"
+
+using namespace LAMMPS_NS;
+using namespace FixConst;
+
+/****************************************************************************/
+
+
+/***************************************************************
+ * create class and parse arguments in LAMMPS script. Syntax:
+ * fix ID group-ID mdi_engine [couple <group-ID>]
+ ***************************************************************/
+FixMDIEngine::FixMDIEngine(LAMMPS *lmp, int narg, char **arg) :
+  Fix(lmp, narg, arg),
+  id_pe(NULL), pe(NULL),
+  id_ke(NULL), ke(NULL)
+{
+
+  if (narg > 3)
+    error->all(FLERR,"Illegal fix mdi command");
+
+  // allocate arrays
+  memory->create(add_force,3*atom->natoms,"mdi:add_force");
+  for (int i=0; i< 3*atom->natoms; i++) {
+    add_force[i] = 0.0;
+  }
+
+  // create a new compute pe style
+  // id = fix-ID + pe, compute group = all
+
+  master = (comm->me==0) ? 1 : 0;
+
+  // create instance of the Irregular class
+  irregular = new Irregular(lmp);
+
+  most_recent_init = 0;
+  exit_flag = false;
+  local_exit_flag = false;
+  target_command = new char[MDI_COMMAND_LENGTH+1];
+  command = new char[MDI_COMMAND_LENGTH+1];
+  current_node = new char[MDI_COMMAND_LENGTH];
+  target_node = new char[MDI_COMMAND_LENGTH];
+  strncpy(target_node, "\0", MDI_COMMAND_LENGTH);
+  strncpy(current_node, "@DEFAULT", MDI_COMMAND_LENGTH);
+
+  // create and add a compute for PE calculation
```

Comment:
This code can be much simplified using the overloaded convenience version of `Modify::add_compute()`.